sentry-core/src/main/java/io/sentry/transport/HttpTransport.java

```diff
@@ -0,0 +1,192 @@
+package io.sentry.transport;
+
+import static io.sentry.SentryLevel.DEBUG;
+import static io.sentry.SentryLevel.ERROR;
+
+import io.sentry.ISerializer;
+import io.sentry.SentryEvent;
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.util.Nullable;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.Proxy;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.Charset;
+import javax.net.ssl.HttpsURLConnection;
+
+/**
+ * An implementation of the {@link ITransport} interface that sends the events to the Sentry server
+ * over HTTP(S) in UTF-8 encoding.
+ */
+public class HttpTransport implements ITransport {
+  public static final int HTTP_TOO_MANY_REQUESTS = 429;
+
+  private static final Charset UTF_8 = Charset.forName("UTF-8");
+
+  @Nullable private final Proxy proxy;
+  private final IConnectionConfigurator connectionConfigurator;
+  private final ISerializer serializer;
+  private final int connectionTimeout;
+  private final int readTimeout;
+  private final boolean bypassSecurity;
+  private final URL sentryUrl;
+  private final SentryOptions options;
+
+  /**
+   * Constructs a new HTTP transport instance. Notably, the provided {@code requestUpdater} must set
+   * the appropriate content encoding header for the {@link io.sentry.ISerializer} instance obtained
+   * from the options.
+   *
+   * @param options sentry options to read the config from
+   * @param proxy the proxy to use, if any
+   * @param connectionConfigurator this consumer is given a chance to set up the request before it
+   *     is sent
+   * @param connectionTimeout connection timeout
+   * @param readTimeout read timeout
+   * @param bypassSecurity whether to ignore TLS errors
+   * @throws URISyntaxException when options contain invalid DSN
+   * @throws MalformedURLException when options contain invalid DSN
+   */
+  public HttpTransport(
+      SentryOptions options,
+      @Nullable Proxy proxy,
+      IConnectionConfigurator connectionConfigurator,
+      int connectionTimeout,
+      int readTimeout,
+      boolean bypassSecurity)
+      throws URISyntaxException, MalformedURLException {
+    this.proxy = proxy;
+    this.connectionConfigurator = connectionConfigurator;
+    this.serializer = options.getSerializer();
+    this.connectionTimeout = connectionTimeout;
+    this.readTimeout = readTimeout;
+    this.options = options;
+    this.sentryUrl = new URI(options.getDsn()).toURL();
+    this.bypassSecurity = bypassSecurity;
+  }
+
+  // visible for testing
```

Comment:
whats about creating an annotation for it? we might need in many other places, like
https://android.googlesource.com/platform/frameworks/support/+/ee8deb2/annotations/src/android/support/annotation/VisibleForTesting.java

Reply:
we already have 2 `io.sentry.util.NonNull`

Reply:
Does it work for `java-library` or requires `android-library`? I say that because this is on `sentry-core`

Reply:
I think the main reason for having them (without an annotation processor) is that the IDE (or at least Intellij) can be set up to take advantage of the annos when offering intentions (at least for `@Nullable` and `@NonNull`). The `@VisibleForTesting` (originally from Guava) is/would be a documentation-only annotation, although there are plugins for e.g. Findbugs that make use of it. As far as I looked Intellij doesn't support it (the "make access weaker" intention has no possibility to be affected by an annotation).

Reply:
Note that I added `@VisibleForTesting` in this PR.